Refactor internal error handling to use exceptions

MikeGitb opened the pull request:
Extracted the exception based error handling from #48 and put it in this separate PR.

Disclaimer: I didn't run any tests on this other than that it compiles (I don't have a hardware device)

daniel-j-h commented on libsweep/include/Error.hpp:
```diff
@@ -0,0 +1,12 @@
+#include <stdexcept>
+
+namespace sweep {
+    struct ErrorBase : std::runtime_error {
```
Can you please lowercase the file to `error.h` as well as `error` here: it's already in the `sweep` namespace, so it's the generic `sweep::error` type.

MikeGitb replied:
OK on the lower case, but removing `Base` will let it collide with the error defined in sweep.cc.

daniel-j-h commented on libsweep/src/protocol.cc:
```diff
@@ -2,10 +2,15 @@
 namespace sweep {
 namespace protocol {
 
+struct Error : ::sweep::ErrorBase {
```
Same here

daniel-j-h commented on libsweep/src/protocol.cc:
```diff
@@ -2,10 +2,15 @@
 #include <stdlib.h>
 
 #include "protocol.h"
+#include "../include/Error.hpp"
```
Please no relative include paths, the include dir should be on the search path already (check CMakeLists.txt)

MikeGitb replied:
Can do.
Although I'm really no fan of unqualified includes for files outside the same directory - in particular for files with very generic names.

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -1,15 +1,18 @@
 #include "sweep.h"
 #include "protocol.h"
 #include "serial.h"
+#include "../include/Error.hpp"
```
Same here

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -19,25 +22,17 @@ typedef struct sweep_device {
 #define SWEEP_MAX_SAMPLES 4096
 
-typedef struct sweep_scan {
+struct sweep_scan {
```
Hm if you want to remove typedefs feel free to do it for all structs

MikeGitb replied:
Must have slipped through, will fix. 
I originally had all the typedefs removed throughout the library but wanted to keep this PR clean of it (there is already going on quite a lot anyway), so I reverted those changes - except that I apparently overlooked this one.

MikeGitb replied:
Actually I think I'll just leave it for now and provide another PR once / if this one gets merged

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
+  using ::sweep::ErrorBase::ErrorBase;
+};
+
+// forward declarations for internal definitions
```
Why? We include the `sweep.h` header anyway, no?

MikeGitb replied:
See comment at the end

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
+
+struct Error : ::sweep::ErrorBase {
+  using ::sweep::ErrorBase::ErrorBase;
+};
```
Shouldn't we use the base from error.h here?

MikeGitb replied:
That would also be a possibility. 
In general I wanted to distinguish between exceptions thrown in the library in general and exceptions that are thrown on this particular level (e.g. as opposed to the IO or protocol level). But currently there is admittedly nothing that makes use of that distinction. We could even use the same exception type everywhere in the code.

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
 
-  sweep_error_s stoperror = nullptr;
-  sweep_device_stop_scanning(out, &stoperror);
+  auto out = std::unique_ptr<sweep_device>(new sweep_device{serial, /*is_scanning=*/true});
```
Why the unique ptr here? For cleaning up in case of exceptions? This is a bit ugly

MikeGitb replied:
Precisely. `sweep_device_stop_scanning` can throw and besides it is my policy to not have a naked `new` anywhere in my code if I can at all help it. 

What would be your suggestion (in c++14 you can use make_unique)?

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
 }
 
-void sweep_device_destruct(sweep_device_s device) {
+void sweep_device_destruct_impl(sweep_device_s device) {
```
Why the change to _impl here?

MikeGitb replied:
To avoid naming collision (the API I version doesn't hav an error parameter to distinguish between the two). See comment below about why I have two in the first place

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -212,7 +160,7 @@ sweep_scan_s sweep_device_get_scan(sweep_device_s device, sweep_error_s* error)
   }
 
-  auto out = new sweep_scan;
+  auto out = std::unique_ptr<sweep_scan>(new sweep_scan);
```
Same here

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
-    *error = sweep_error_construct("unable to receive sample rate command response");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
+template <class F> auto translateException(F&& f, sweep_error_s* error) -> decltype(std::forward<F>(f)()) {
```
The following should probably be in the error.h header

MikeGitb replied:
As it is translating an exception into and ABI error, which is defined in `sweep.h` and no other part of the code has to care about this I don't see why this should be moved there.

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
+    return std::forward<F>(f)();
+  } catch (const sweep::ErrorBase& e) {
+    *error = new sweep_error{e.what()};
+  } catch (const std::bad_alloc&) {
```
Why not catching `std::exception&` for which there is a `.what()`

MikeGitb replied:
Overlooked, will fix

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
+  return {};
+}
+
+template <class F> void translateExceptionV(F&& f, sweep_error_s* error) {
```
Would be great if we can overload this function instead

MikeGitb replied:
As I said in the other thread: If you see a simple way to do this be my guest. also I really don't see what the advantage would be.

daniel-j-h commented on libsweep/src/sweep.cc:
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
 }
+} // local namespace
+
+/* ###################### implementation of public API ################################# */
```
Ah now I understand - you split off the C API and the internal functions. Hm I see no benefit in doing so, what's the reason behind this?

MikeGitb replied:
Separation of concerns: The private functions are concerned with implementing the functionality the implementation of the public functions then only needs to take care about eror translation and any other adaptions to the ABI that might be necessary in the future. Also `translateException` needs some function / function object to be passed to it anyway, so the only alternative I see would be to wrapp the internal functionality into a lambda which doesn't really make the code more readable.

daniel-j-h commented on libsweep/src/unix/serial.cc:
```diff
@@ -1,6 +1,7 @@
 #define _POSIX_C_SOURCE 200809L
 
 #include "serial.h"
+#include "../../include/Error.hpp"
```
Same here

MikeGitb commented:
@daniel-j-h: Ok, I
- made `error.hpp` and `error` lower case
- changed to unqualified includes
- changed the error translation a bit

I did not incorporate the rest of your feedback and am waiting for your responses to my comments instead. I'm not sure, why git marks them all as outdated

MikeGitb commented:
Rebased onto master

MikeGitb commented:
Rebased on master and rewrote error handling in win/serial.cc to use exceptions

MikeGitb commented:
So ist there any chance this will get merged in the near future?

daniel-j-h commented:
Yes, please excuse my inactivity I'm out until late next week. Maybe
@dcyoung can jump in here - otherwise you will have to wait a bit. I
want to get us to switch to exceptions and references for sure.


On 03/05/2017 12:08 AM, MikeGitb wrote:
> So ist there any chance this will get merged in the near future?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scanse/sweep-sdk/pull/59#issuecomment-284191550>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgLiYC1yaPcQRpsUymHY9bNDE0_xZb-ks5rie71gaJpZM4MFTU8>.
>

MikeGitb commented:
No rush. Main question is, if you are OK with the general structure of this PR as it is now.

MikeGitb commented:
Closing this for now - I think after the last round of changes a merge is no longer possible. I won't delete the branch though, as parts might be salvageable for future PRs
